### ISSUE TITLE
systemd-resolved's local DNS stub incorrectly returs NOTIMP on DO/CD queries

### ISSUE THREAD

**Issue by letoams** [systemd/systemd#17218]
systemd-246.4-2.fc33.x86_64

When sending a query with the DO and CD flags set, the DNS reply is NOTIMP

```
paul@thinkpad:~/$ dig +dnssec nohats.ca @127.0.0.53 +cd

; <<>> DiG 9.11.22-RedHat-9.11.22-1.fc33 <<>> +dnssec nohats.ca @127.0.0.53 +cd
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOTIMP, id: 6474
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1
```

It is wrong to return NOTIMP. This will just direct the DNS client to try another nameserver from /etc/resolv.conf, or give up if there are no more others. NOTIMP must only be returned for not implemented OPCODEs.

When you don't support a certain flag, you are supposed to ignore it. So when +do +cd is set, you return the answer without do /cd bit in the OPT record. And obviously without the ad bit. If not understanding the flag means something did not get returned, so be it. That is fine. If it means you have nothing to return, you send NODATA or NXDOMAIN. Sending NOTIMP violates the DNS specification.

**Comment by poettering** [comment]
Hmm, RFC1035 is not that clear about what NOTIMP means, just says "The name server does not support the requested kind of query". Which is pretty vague, there's no direct reference to the opcode field mentioned a bit further up. Is there a later RFC that clarifies that?

I mean, we can certainly change what we return if we don#t support something, but I am not sure how much this would do good... DNSSEC clients can barely rely on DNS servers to return useful data on success, but as long as there's an error returned at all it's already excellent compared to much of the DNS servers I tested against.

Note that the dns stub in resolved actually does return some DNSSEC metadata on DO/non-CD requests: we include the amount of info some well-known clients actually use. Specifically glibc's nss-dns looks for the AD flag, so we set that, but don't bother with more data (since the fewest deployed clients actually do DNSSEC checking client side). Yes, we are not implementing DNSSEC towards the client beyond some basics, it was not supposed to so far.

(Also note that we take the liberty to set AD for some data from local trusted sources, for example for records we synthesize from /etc/hosts, as that's a source under control of the local admin, which we generally trust the most, i.e. more than than the public internet DNS. I figure to some DNSSEC purists this migh be revolting...)

**Comment by pemensik** [comment]
> Hmm, RFC1035 is not that clear about what NOTIMP means, just says "The name server does not support the requested kind of query". Which is pretty vague, there's no direct reference to the opcode field mentioned a bit further up. Is there a later RFC that clarifies that?
Requested kind of query means OPCODE in RFC 1035. It does not mean queried type (QTYPE) or rrset in question section. As a forwarder, it should forward and cache even for records it does not know nothing about. It does not have to understand all records requested.
> 
> Note that the dns stub in resolved actually does return some DNSSEC metadata on DO/non-CD requests: we include the amount of info some well-known clients actually use. Specifically glibc's nss-dns looks for the AD flag, so we set that, but don't bother with more data (since the fewest deployed clients actually do DNSSEC checking client side). Yes, we are not implementing DNSSEC towards the client beyond some basics, it was not supposed to so far.

This approach is forward incompatible. You used only bits you know are used now by clients. If you act as forwarder, you should cache and forward original response. If systemd supports DNSSEC (it does!), it should just remove DNSSEC related bits from reply, if original client did not set DO flag. That is rrsig, nsec, nsec3 would not be delivered. DNS evolves and your version should be able work well even after 5 years, especially if supported in RHEL.

Dnsmasq for example stores original reply message and only adjusts query id on reply. It workaround inability to strip DNSSEC records to client. Instead, it makes queries with DO bit set bypass cache and always forwards. That way, it can cache non-dnssec enabled replies and not collide with DO enabled query for cached entry without DO. I doubt this is correct approach. Solid DNS implementations always use DO on outgoing queries, just filter DNSSEC rrsets on reply to DO unaware client.

```
# cache non-DO
dig @127.0.0.53 fedoraproject.org
# retry with DO, must contain also RRSIG
dig +dnssec @127.0.0.53 fedoraproject.org
```
> 
> (Also note that we take the liberty to set AD for some data from local trusted sources, for example for records we synthesize from /etc/hosts, as that's a source under control of the local admin, which we generally trust the most, i.e. more than than the public internet DNS. I figure to some DNSSEC purists this migh be revolting...)
You are misusing [AD flag](https://tools.ietf.org/html/rfc4035#section-3.2.3) for something else. AA flag from RFC 1035 should be set for any authoritative records instead. Because they are just authoritative, but otherwise unverified. Check dig localhost against any DNS server and flags.

**Comment by letoams** [comment]
> Hmm, RFC1035 is not that clear about what NOTIMP means, just says "The name server does not support the requested kind of query". Which is pretty vague, there's no direct reference to the opcode field mentioned a bit further up. Is there a later RFC that clarifies that?

RFC 1035 predates DNSSEC. There have been updates on this, eg RFC 3225 explains when to reply with the DO bit.

RFC 8906 clarifies when to use FORMERR vs NOTIMP - It suggests you may return FORMERR.

RFC 9806 Section 8.1.3.1. Testing CD=1 Queries, clarifies the behaviour to ignore if DNSSEC is not supported and copy if it is. 

That is, neither FORMERR or NOTIMP should be returned when receiving the +CD bit.

Technically, the problem now is what "DNSSEC support" means for systemd-resolved. With its current behaviour of returning cached answers without DNSSEC data in response to +DO queries, it cannot claim DNSSEC support. By doing so, it will will appear to DNSSEC capable clients of being malicious or compromised.

It would be best if the incoming CD bit is ignored and not copied to the reply, and to return a non-DNSSEC answer without DO and CD bits set. That makes it clear to DNSSEC capable clients that this server cannot be used for DNSSEC queries and in theory it can try another server listed in resolv.conf.

**Comment by pemensik** [comment]
This request seems to match much older issue #4621, which is still unresolved.

**Comment by poettering** [comment]
> Requested kind of query means OPCODE in RFC 1035. It does not mean queried type (QTYPE) or rrset in question section.

The text in the RFC is pretty vague, not sure where you get your understanding from. I doubt it's that clear cut. Is there a clarification RFC for what you are claiming?

**Comment by letoams** [comment]
My comment above shared pointers to clarifications about old DNS RFCs including RFC 1035. Did you not find the clarifications you were looking for in these subsequent documents? We agree that RFC 1035 is "pretty vague" and "not clear cut". But I thought that the RFCs cited above that update RFC 1035 were clear?

For example: RFC 8906 Section 3.1.2 clarifies:

3.1.2.  Unknown/Unsupported Type Queries

   Some servers fail to respond to unknown or unsupported types.  If a
   server receives a query for a type that it doesn't recognise, or
   doesn't implement, it is expected to return the appropriate response
   as if it did recognise the type but does not have any data for that
   type, i.e., either NOERROR or NXDOMAIN.  The exceptions to this are
   queries for Meta-RR types, which may return NOTIMP.

and section 3.1.4:

3.1.4.  Unknown DNS Opcodes

   The use of previously undefined opcodes is to be expected.  Since the
   DNS was first defined, two new opcodes have been added, UPDATE and
   NOTIFY.

   NOTIMP is the expected rcode to an unknown or unimplemented opcode.


I think these clarifications should be interpreted to mean that receiving qname/qtype queries that are known, with DO bit that are either "not supported or not configured", should not return NOTIMP, but should return the query data without DO/AD bit set.

RFC 9806 further clarifies the use of the CD bit in that it should be ignored in the query if not supported or configured. It does not state that NOTIMP is a valid response (and RFC 8906 states why returning NOTIMP is generally bad and should only be done for unknown OPCODEs)

If you have questions about the clarifications offered in RFC 8906 and 9806, please let me know.

**Comment by pemensik** [comment]
> > Requested kind of query means OPCODE in RFC 1035. It does not mean queried type (QTYPE) or rrset in question section.
> 
> The text in the RFC is pretty vague, not sure where you get your understanding from. I doubt it's that clear cut. Is there a clarification RFC for what you are claiming?

https://tools.ietf.org/html/rfc1035#section-4.1.1

>OPCODE          A four bit field that specifies **kind of query** in this
>                message.  This value is set by the originator of a query
>                and copied into the response. 
>...
>RCODE           Response code - this 4 bit field is set as part of
>                responses.  The values have the following
>                interpretation:
>
>                0               No error condition
>
>                1               Format error - The name server was
>                                unable to interpret the query.
>
>                2               Server failure - The name server was
>                                unable to process this query due to a
>                                problem with the name server.
>
>                3               Name Error - Meaningful only for
>                                responses from an authoritative name
>                                server, this code signifies that the
>                                domain name referenced in the query does
>                                not exist.
>
>                4               Not Implemented - The name server does
>                                not support the requested *kind of query*.
>
>                5               Refused - The name server refuses to
>                                perform the specified operation for
>                                policy reasons.  For example, a name
>                                server may not wish to provide the
>                                information to the particular requester,
>                                or a name server may not wish to perform
>                                a particular operation (e.g., zone

If you are ever unsure, how it should behave, multiple implementors and operators of DNS services would help you at [DNS-OARC mailing list](https://lists.dns-oarc.net/mailman/listinfo/dns-operations)

**Comment by loganaden** [comment]
> 
> (Also note that we take the liberty to set AD for some data from local trusted sources, for example for records we synthesize from /etc/hosts, as that's a source under control of the local admin, which we generally trust the most, i.e. more than than the public internet DNS. I figure to some DNSSEC purists this migh be revolting...)

Is there a way to disable this behaviour as an operator ?

**Comment by pemensik** [comment]
I am not sure what should it change. But it seems still broken. Only less a bit.

When CD flag is included, it seems it provides also RRSIG records:
```
$ dig +dnssec +cd @127.0.0.53

; <<>> DiG 9.16.8-RedHat-9.16.8-1.fc32 <<>> +dnssec +cd @127.0.0.53
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 25848
;; flags: qr rd ra cd; QUERY: 1, ANSWER: 14, AUTHORITY: 0, ADDITIONAL: 27

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags: do; udp: 65494
;; QUESTION SECTION:
;.				IN	NS

;; ANSWER SECTION:
.			94322	IN	NS	a.root-servers.net.
.			94322	IN	NS	i.root-servers.net.
.			94322	IN	NS	g.root-servers.net.
.			94322	IN	NS	d.root-servers.net.
.			94322	IN	NS	f.root-servers.net.
.			94322	IN	NS	l.root-servers.net.
.			94322	IN	NS	h.root-servers.net.
.			94322	IN	NS	m.root-servers.net.
.			94322	IN	NS	b.root-servers.net.
.			94322	IN	NS	e.root-servers.net.
.			94322	IN	NS	j.root-servers.net.
.			94322	IN	NS	c.root-servers.net.
.			94322	IN	NS	k.root-servers.net.
.			506852	IN	RRSIG	NS 8 0 518400 20201124050000 20201111040000 26116 . Ap1dCKj6+PjWpUFO2woOgmrXuUlcruV1+2l3d6+qYt51TCjdbCwP0RKB X6OXr14e7ugq0gH3QKBlX9yjzz1X1Xl7BQKQNp3KRG7crz3ImSdE+1Gv SZwpWmAFqj2wsZzy6JQRZyZ+zZYjRjMSD3d+n2cFA/0DhckHvBo+MR3c ygbn3WhgumumX82N0Q9rxPksB3Uh+6Krks4VrxSU9RdDliI1oVwscz9T ZZbkHiioE2QJc7Sc8D6BfMZ6jbNZYhWv4X7IhpwEFbKfYjxkU5Ebix+C NmvPS8shXvUVIe+KsjjrC95S1E6AGtxSbKEgq6cCEqKpvazVXsFAhE8i i/LPYA==

;; ADDITIONAL SECTION:
a.root-servers.net.	136733	IN	A	198.41.0.4
a.root-servers.net.	161253	IN	AAAA	2001:503:ba3e::2:30
b.root-servers.net.	126178	IN	A	199.9.14.201
b.root-servers.net.	161253	IN	AAAA	2001:500:200::b
c.root-servers.net.	161253	IN	A	192.33.4.12
c.root-servers.net.	161253	IN	AAAA	2001:500:2::c
d.root-servers.net.	161253	IN	A	199.7.91.13
d.root-servers.net.	161253	IN	AAAA	2001:500:2d::d
e.root-servers.net.	161253	IN	A	192.203.230.10
e.root-servers.net.	161253	IN	AAAA	2001:500:a8::e
f.root-servers.net.	161253	IN	A	192.5.5.241
f.root-servers.net.	161253	IN	AAAA	2001:500:2f::f
g.root-servers.net.	161253	IN	A	192.112.36.4
g.root-servers.net.	161253	IN	AAAA	2001:500:12::d0d
h.root-servers.net.	161253	IN	A	198.97.190.53
h.root-servers.net.	161253	IN	AAAA	2001:500:1::53
i.root-servers.net.	161253	IN	A	192.36.148.17
i.root-servers.net.	161253	IN	AAAA	2001:7fe::53
j.root-servers.net.	161253	IN	A	192.58.128.30
j.root-servers.net.	161253	IN	AAAA	2001:503:c27::2:30
k.root-servers.net.	161253	IN	A	193.0.14.129
k.root-servers.net.	161253	IN	AAAA	2001:7fd::1
l.root-servers.net.	161253	IN	A	199.7.83.42
l.root-servers.net.	161253	IN	AAAA	2001:500:9f::42
m.root-servers.net.	246938	IN	A	202.12.27.33
m.root-servers.net.	593253	IN	AAAA	2001:dc3::35

;; Query time: 0 msec
;; SERVER: 127.0.0.53#53(127.0.0.53)
;; WHEN: St lis 11 21:00:12 CET 2020
;; MSG SIZE  rcvd: 1097
```

But when only DO flag is requested, no signature is received.
```
$ dig +dnssec @127.0.0.53

; <<>> DiG 9.16.8-RedHat-9.16.8-1.fc32 <<>> +dnssec @127.0.0.53
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 63115
;; flags: qr rd ra ad; QUERY: 1, ANSWER: 13, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags: do; udp: 65494
;; QUESTION SECTION:
;.				IN	NS

;; ANSWER SECTION:
.			7021	IN	NS	c.root-servers.net.
.			7021	IN	NS	h.root-servers.net.
.			7021	IN	NS	g.root-servers.net.
.			7021	IN	NS	m.root-servers.net.
.			7021	IN	NS	f.root-servers.net.
.			7021	IN	NS	a.root-servers.net.
.			7021	IN	NS	j.root-servers.net.
.			7021	IN	NS	e.root-servers.net.
.			7021	IN	NS	i.root-servers.net.
.			7021	IN	NS	k.root-servers.net.
.			7021	IN	NS	l.root-servers.net.
.			7021	IN	NS	d.root-servers.net.
.			7021	IN	NS	b.root-servers.net.

;; Query time: 0 msec
;; SERVER: 127.0.0.53#53(127.0.0.53)
;; WHEN: St lis 11 21:01:38 CET 2020
;; MSG SIZE  rcvd: 239
```

```
$ delv +cd @127.0.0.53
; fully validated
.			94203	IN	NS	a.root-servers.net.
.			94203	IN	NS	i.root-servers.net.
.			94203	IN	NS	g.root-servers.net.
.			94203	IN	NS	d.root-servers.net.
.			94203	IN	NS	f.root-servers.net.
.			94203	IN	NS	l.root-servers.net.
.			94203	IN	NS	h.root-servers.net.
.			94203	IN	NS	m.root-servers.net.
.			94203	IN	NS	b.root-servers.net.
.			94203	IN	NS	e.root-servers.net.
.			94203	IN	NS	j.root-servers.net.
.			94203	IN	NS	c.root-servers.net.
.			94203	IN	NS	k.root-servers.net.
.			94203	IN	RRSIG	NS 8 0 518400 20201124050000 20201111040000 26116 . Ap1dCKj6+PjWpUFO2woOgmrXuUlcruV1+2l3d6+qYt51TCjdbCwP0RKB X6OXr14e7ugq0gH3QKBlX9yjzz1X1Xl7BQKQNp3KRG7crz3ImSdE+1Gv SZwpWmAFqj2wsZzy6JQRZyZ+zZYjRjMSD3d+n2cFA/0DhckHvBo+MR3c ygbn3WhgumumX82N0Q9rxPksB3Uh+6Krks4VrxSU9RdDliI1oVwscz9T ZZbkHiioE2QJc7Sc8D6BfMZ6jbNZYhWv4X7IhpwEFbKfYjxkU5Ebix+C NmvPS8shXvUVIe+KsjjrC95S1E6AGtxSbKEgq6cCEqKpvazVXsFAhE8i i/LPYA==
```

```
$ delv @127.0.0.53
;; validating ./NS: got insecure response; parent indicates it should be secure
;; insecurity proof failed resolving './NS/IN': 127.0.0.53#53
;; resolution failed: insecurity proof failed
```

```
build/systemd-resolved --version
systemd 247 (247-rc1-211-gcc8f745)
+PAM +AUDIT +SELINUX +IMA -APPARMOR +SMACK +SYSVINIT +UTMP -LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ -LZ4 +ZSTD -SECCOMP +BLKID +ELFUTILS -KMOD +IDN2 -IDN +PCRE2 default-hierarchy=unified
```

Commit cc8f745bfcd35169e7eaec079268119068fa2886

**Comment by pemensik** [comment]
Almost there, thank you for the progress.

```
$ delv @127.0.0.53 +cd id.fedoraproject.org | head -1
; fully validated
```

**Comment by poettering** [comment]
Fixed by 30ee7071703226bf84e69f983ad1c08283e4b891 and 775ae35403f8f3c01b7ac13387fe8aac1759993f
